Add Custom Email Backend

Jtang-1 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Builds on https://github.com/dimagi/commcare-hq/pull/33518

This allows domains to configure a custom gateway with AWS SES such that conditional alert and broadcast emails can be sent with a custom email domain. It'll require setup on AWS SES and SNS.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Jira Ticket: https://dimagi-dev.atlassian.net/browse/USH-3467
https://dimagi-dev.atlassian.net/browse/USH-3468
Spec: https://docs.google.com/document/d/1HjW8xpztIZYPLKRyp6C1F-f66kwNNO0r1JaR6InnUVE/

If a custom `EmailSetting` is configured on the domain, an `EmailBackend` [with a different SMTP configuration](https://docs.djangoproject.com/en/4.2/topics/email/#smtp-backend) is instantiated based on the domain's `EmailSetting`. This custom instance of `EmailBackend` is passed as the `connection` to use for `EmailMessage` as described [here](https://docs.djangoproject.com/en/4.2/topics/email/#django.core.mail.EmailMessage:~:text=class%20EmailMessage-,%C2%B6,-The%20EmailMessage%20class). This connection is used to access AWS SES and the `from_email` is a verified identity on AWS SES that will be used as the email's sender.

In relation to email tracking:
`X- SES-CONFIGURATION-SET` header uses the user defined configuration set if defined and `use_tracking_headers` is enabled. 
`X-COMMCAREHQ-MESSAGE-ID` header is also used for email tracking to tell HQ which MessagingSubEvent to update when receiving the notification from SES. It is only relevant to email tracking if a SES configuration set is used.  However, this header will be added if available regardless of `use_tracking_headers` toggle.

See related docs to setting up AWS SES [here](https://confluence.dimagi.com/display/USH/Allow+user+to+define+custom+email+gateway+that+can+be+used+to+send+emails+from+HQ).

As of this PR, only emails sent through `EmailContent` (such as conditional alerts and broadcasts) will use the custom domain gateway (It uses `CustomEmailConfiguration` if configured, otherwise defaults to `DefaultEmailConfiguration`). All other email types will continue to use the default `from_email` but it is planned for system emails to also support using the custom domain 
gateway.

For reviewers: 
There was a change in design decision to support only one gateway and the need to encapsulate `from_email` and `connection`. Commit [bfb6279](https://github.com/dimagi/commcare-hq/pull/33484/commits/bfb6279cfc41bd0d5ad2bfaf71dfebdc9563ec49) is the changing point.
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
new feature flag - custom_email_gateway

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Locally tested and on staging. It defaults to locally defined settings that were not changed as part of this PR so scope is limited to domains with custom email backend.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Will QA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
adds custom interface for Django's email get_connection

Jtang-1 committed:
gets connection and passes to EmailMessage

Jtang-1 committed:
get domain custom backend if configured and instantiate backend

esoergel commented on corehq/apps/hqwebapp/tasks.py:
```diff
@@ -93,6 +95,7 @@ def send_mail_async(self, subject, message, from_email, recipient_list,
             from_email=from_email,
```
I just realized this will of course need to be part of the configuration

Jtang-1 replied:
yes it does, I added it here [e70be68](https://github.com/dimagi/commcare-hq/pull/33484/commits/e70be68dedee29ab617569ecf239f513145c16b3) and use it here for `send_mail_async`

esoergel commented on corehq/apps/hqwebapp/tasks.py:
```diff
@@ -93,6 +95,7 @@ def send_mail_async(self, subject, message, from_email, recipient_list,
             from_email=from_email,
             to=filtered_recipient_list,
             headers=headers,
```
Can you investigate these headers?  How many of them are gateway specific?  I wonder whether `connection` encapsulates enough of this infrastructure.  I think `from_email` and `headers` will likely need to be part of what's provided by the custom backend.

Jtang-1 replied:
I looked into it, there's 3 other headers:
- `X-COMMCAREHQ-MESSAGE-ID` This is the id of the `MessagingSubEvent` item so wouldn't be specific to the custom backend
- `Return-Path` This is the return email used to forward BOUNCE & COMPLAINT notifications. This would be specific to the custom backend if partners wanted to monitor their bounces/complaints and want a custom email for this. For now, I don't think that's needed
- `X-SES-CONFIGURATION-SET` - This would be specific to the custom backend. So that SNS can probably reach our log_email_event endpoint. Still in the process of figuring out how this works but the SNS endpoint secret field that Akash added should be all that's needed.

esoergel replied:
Thanks, that's helpful.  I wasn't familiar with `Return-Path`, but [I found this article on it](https://sendgrid.com/blog/what-is-return-path/).  Sounds like it's important in establishing credibility of the email sender.  I agree that we can leave `X-COMMCAREHQ-MESSAGE-ID` in for everything.  It should be fine to get email working end-to-end without the other stuff for a first pass, though I think we will need to get a configurable `Return-Path` set up eventually, as well as the SNS stuff.

Jtang-1 commented:
Rebased base branch to be https://github.com/dimagi/commcare-hq/pull/33518 instead of https://github.com/dimagi/commcare-hq/pull/33386

Jtang-1 committed:
use new EmailSettings model instead of supporting multiple backend cofigurations

Jtang-1 committed:
bug: from_address is not hooked up with new connection

Jtang-1 committed:
creates an object with from_address and connection since they are binded

Jtang-1 committed:
passes from_email and connection as object to send_mail_async

This is done in preparation to support both the default and custom email connection

snopoke commented on corehq/apps/hqwebapp/tasks.py:
```diff
@@ -43,8 +44,8 @@ def mark_subevent_gateway_error(messaging_event_id, error, retrying=False):
       bind=True, default_retry_delay=15 * 60, max_retries=10, acks_late=True)
-def send_mail_async(self, subject, message, from_email, recipient_list,
-                    messaging_event_id=None, domain=None):
+def send_mail_async(self, subject, message, configuration: EmailConfigurationManager,
+                    recipient_list, messaging_event_id=None, domain: str = None):
```
I think our current policy to not to use type hints

Jtang-1 replied:
I didn't know that, I personally find it helpful but I would be fine with removing it. The only case is for `configuration` -is it more standard to just add a comment that `configuration` is a `EmailConfigurationManager`? Otherwise, it'll be hard to figure out just by reading the code.

Jtang-1 committed:
uses custom configured SNS configuration set and SNS secret for log_email_event

Jtang-1 commented:
This PR was in a weird place and kept getting changes as I was figuring things through - but it is now ready for review. It implements the basics required to email end-to-end and also monitoring email events. It does not implement a custom `Return-Path` header

esoergel commented on corehq/ex-submodules/dimagi/utils/django/email.py:
```diff
@@ -168,3 +169,75 @@ def send_HTML_email(subject, recipient, html_content, text_content=None,
             error_msg.send()
+
+
+def getEmailConfiguration(domain: str, from_email: str = settings.DEFAULT_FROM_EMAIL):
```
why the camelCase?

Jtang-1 replied:
oops [a0a399c](https://github.com/dimagi/commcare-hq/pull/33484/commits/a0a399cbcdab3671bff35843a360912c82d1b7c1)

esoergel commented on corehq/ex-submodules/dimagi/utils/django/email.py:
```diff
@@ -168,3 +169,75 @@ def send_HTML_email(subject, recipient, html_content, text_content=None,
+def getEmailConfiguration(domain: str, from_email: str = settings.DEFAULT_FROM_EMAIL):
+    from corehq.apps.email.models import EmailSettings
+    try:
+        email_setting = EmailSettings.objects.get(domain=domain)
+        if email_setting.use_this_gateway:
+            return CustomEmailConfiguration(email_setting)
+        else:
+            return DefaultEmailConfiguration(from_email)
```
```suggestion
        email_setting = EmailSettings.objects.get(domain=domain, use_this_gateway=True)
        return CustomEmailConfiguration(email_setting)
```

Jtang-1 replied:
Is that equivalent? email_setting would be empty if `use_this_gateway` is False and it wouldn't return `DefaultEmailConfiguration`

esoergel replied:
It just moves the check to the DB lookup - if there's nothing on that domain with `use_this_gateway=True`, it'll raise `DoesNotExist` and return the default after the `except`

Jtang-1 replied:
you're right, i forgot i had that enclosed in a try block [9cf170c](https://github.com/dimagi/commcare-hq/pull/33484/commits/9cf170cd8c05e9c4a05696c221cf626172557c53)

esoergel commented on corehq/ex-submodules/dimagi/utils/django/email.py:
```diff
@@ -168,3 +169,75 @@ def send_HTML_email(subject, recipient, html_content, text_content=None,
+        return DefaultEmailConfiguration(from_email)
+
+
+class EmailConfigurationManager(ABC):
```
:star:

esoergel reviewed (commented):
nicely done!

Jtang-1 commented on corehq/apps/hqwebapp/tasks.py:
```diff
@@ -43,8 +44,8 @@ def mark_subevent_gateway_error(messaging_event_id, error, retrying=False):
       bind=True, default_retry_delay=15 * 60, max_retries=10, acks_late=True)
-def send_mail_async(self, subject, message, from_email, recipient_list,
-                    messaging_event_id=None, domain=None):
+def send_mail_async(self, subject, message, configuration: EmailConfigurationManager,
```
Thanks for pointing out the celery issue - I wasn't aware of that. The only other option I can think of is having the caller decompose the properties and pass the arguments individually to `send_mail_async`. Or is there is a better option?

> It looks like this arg is passed in to allow control over which gateway - why not just switch everything to the custom gateway when configured to the domain? Is it to do a phased rollout or something?

I wanted to limit this PR to only conditional alerts and broadcast as that was the scope of the ticket. I haven't looked into each of the callers of `send_mail_async` to know what it would impact, I plan to do that in the ticket to implement this to all system emails. But I don't think all emails would use the custom gateway. For example, I think emails involved with creating a Commcare account uses `send_mail_async` and we would to use the default Gateway for that.

esoergel replied:
That makes sense - I do think the eventual outcome is that any emails tagged with a domain will use the domain's custom gateway, but it's certainly fair to be cautious about that.

> The only other option I can think of is having the caller decompose the properties and pass the arguments individually to send_mail_async. Or is there is a better option?

Another option is to pass in an arg like `use_domain_gateway`, and then in the body of this function do something like:
```python
if use_domain_gateway and domain_gateway_exists:
    configuration = getEmailConfiguration
else:
    configuration = DefaultEmailConfiguration
```
Nice thing about that is it requires callers of this function to have less knowledge of how this works (and fewer imports)

Jtang-1 replied:
Thanks for the suggestion, i like this method
[37107ae](https://github.com/dimagi/commcare-hq/pull/33484/commits/37107aedbcdf333864073bd984334240d03a48b7)

Jtang-1 committed:
Merge branch 'master' into jt/custom_email_backend

Jtang-1 committed:
refactor: adds use_domain_gateway flag instead of passing in EmailConfiguration into send_mail_async

Jtang-1 committed:
refactor: move check to dblookup

Jtang-1 committed:
style: corrects from camelcase to snakecase

Jtang-1 committed:
feat: doesn't add config set header if domain email setting disables use_tracking_headers

Jtang-1 committed:
lint

Jtang-1 committed:
refactors for testing and bug: passes domain to get_email_configuration

Jtang-1 committed:
tests: adds tests correct emailsettings is uesd and has the correct connection

Jtang-1 committed:
test: update tests to reflect new send_mail_async parameters

Jtang-1 committed:
lint
